File: Java/ZeroSumSubArray.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

public class ZeroSumSubArray {
    public static void main (String[] args) {
        //int  c=  maxLen(new int[] { 15, -2, 2, -8, 1, 7, 10, 23 } , 8);

        int c = findSubarray(new int[] {  0, 0, 5, 5, 0, 0 }, 6 );

        System.out.println(c);
    }

    static int maxLen(int arr[], int n) {
        var sum = 0;
        var count = 0;

        HashMap map = new HashMap<>();

        for (int i = 0; i < n; i++) {
            sum += arr[i];

            if (sum == 0) count = i;

            if(!map.containsKey(sum))
                map.put(sum, i);
            else
                count = Math.max(i - (int)map.get(sum), count);
        }
        return count;
    }


    public static int findSubarray(int[] arr ,int n) 
    {
        int sum = 0;
        int count = 0;

        HashMap<Integer, ArrayList<Integer>> map = new HashMap<Integer, ArrayList<Integer>>();

        for (int i = 0; i < n; i++) {

            sum += arr[i];

            if (sum == 0) count++;

            if (map.keySet().contains(sum)) 
                count += ((ArrayList<Integer>)map.get(sum)).size();
            else
                map.put(sum, new ArrayList<>());


            map.get(sum).add(i + 1);
        }

        return count;
    }

}

File: GeeksForGeeksProblems/ZeroSumSubArray.cs
//https://practice.geeksforgeeks.org/problems/largest-subarray-with-0-sum/1?page=1&company[]=Microsoft&curated[]=1&sortBy=submissions

using System;
using System.Collections.Generic;
using System.Linq;

namespace GeeksForGeeksProblems
{
    public class ZeroSumSubArray
    {
        public static int GetSizeOfLargestZeroSumSubArray(int[] arr)
        {
            var sum = 0;
            var count = 0;

            var dictionary = new Dictionary<int, int>();

            for (int i = 0; i < arr.Length; i++)
            {
                sum += arr[i];

                if (sum == 0) { count = i + 1; continue; }

                if (!dictionary.Keys.Contains(sum))
                    dictionary.Add(sum, i);
                else
                    count = Math.Max(i - dictionary[sum], count);
            }

            return count;
        }

        public static int GetSizeOfLargestZeroSumSubArray1(int[] arr)
        {
            var subArrays = new List<List<int>>();

            for (int i = 0; i < arr.Length; i++)
            {
                for (int j = 0; j < arr.Length; j++)
                {
                    if (arr.Skip(i).Take(j + 1).Sum() == 0)
                    {
                        subArrays.Add(arr.Skip(i).Take(j + 1).ToList());
                    }
                }
            }

            return subArrays.Max(list => list.Count());
        }
    }
}
